pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -421,30 +373,108 @@ func (s *NodeController) checkNodeReady(node *api.Node) *api.NodeCondition {
+				}
+			}
+		}
+	}
+	return nodes, nil
+}
+
+// MonitorNodeStatus verifies node status are constantly updated by kubelet, and if
+// not, post node NotReady status. It also evicts all pods if node is not ready for
+// a long period of time.
+func (s *NodeController) MonitorNodeStatus() error {
+	nodes, err := s.kubeClient.Nodes().List()
 	if err != nil {
 		return err
 	}
-	for _, pod := range pods.Items {
-		if pod.Status.Host != nodeID {
-			continue
+	for i := range nodes.Items {
+		node := &nodes.Items[i]
+		// Precompute all condition times to avoid deep copy of node status (We'll modify node for
+		// updating, and NodeStatus.Conditions is an array, which makes assignment copy not useful).
+		latestConditionTime := s.latestConditionTime(node, api.NodeReady)
+		lastProbeTime := s.getCondition(node, api.NodeReady).LastProbeTime
+		lastTransitionTime := s.getCondition(node, api.NodeReady).LastTransitionTime
+		var gracePeriod time.Duration
+		if latestConditionTime == node.CreationTimestamp {
+			gracePeriod = nodeStartupGracePeriod
+		} else {
+			gracePeriod = nodeMonitorGracePeriod
 		}
-		glog.V(2).Infof("Delete pod %v", pod.Name)
-		if err := s.kubeClient.Pods(pod.Namespace).Delete(pod.Name); err != nil {
-			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
+		// NodeReady condition was last set longer ago than gracePeriod, so update it to ConditionNone
+		// (regardless of its current value) in the master, without contacting kubelet.
+		if util.Now().After(latestConditionTime.Add(gracePeriod)) {
+			readyCondition := s.getCondition(node, api.NodeReady)
+			if readyCondition == nil {
+				glog.V(2).Infof("node %v is never updated by kubelet. latestConditionTime is %v",
+					node.Name, latestConditionTime)
+				node.Status.Conditions = append(node.Status.Conditions, api.NodeCondition{
+					Type:               api.NodeReady,
+					Status:             api.ConditionNone,
+					Reason:             fmt.Sprintf("Kubelet never posted node status"),
+					LastProbeTime:      util.Now(),
+					LastTransitionTime: util.Now(),
+				})
+			} else {
+				// Note here the out-dated condition can be the one posted by nodecontroller
+				// itself before. We keep posting the status to keep LastProbeTime fresh.
+				glog.V(2).Infof("node %v hasn't been updated for %v: out-dated condition is %+v",
+					node.Name, gracePeriod, readyCondition)
+				readyCondition.Status = api.ConditionNone
+				readyCondition.Reason = fmt.Sprintf("Kubelet stopped posting node status")
+				readyCondition.LastProbeTime = util.Now()
+				if readyCondition.Status == api.ConditionFull {
+					readyCondition.LastTransitionTime = util.Now()
+				}
+			}
+			_, err = s.kubeClient.Nodes().Update(node)
+			if err != nil {
+				glog.Errorf("error updating node %s: %v", node.Name, err)
+			}
+		}
+		// Eviction timeout! Evict all pods on the unhealthy node.
+		if lastProbeTime.After(lastTransitionTime.Add(s.podEvictionTimeout)) {
```

Comment:
Why based on lastTransitionTime?

I think probably the logic should be:
if NodeReady is not updated for gracePeriod seconds, transition the node to NodeReady=Unknown
If the node is in NodeReady=Unknown and NodeReady has not been updated for podEvictionTimeout seconds, then evict.

IIUC, that means a node would take gracePeriod+podEvictionTimeout seconds to go from NodeReady=Full or NodeReady=None to pod eviction.

What do you think?